libs/sdk-ui-dashboard/src/_staging/metadata/objRefMap.ts

```diff
@@ -80,28 +91,60 @@ export class ObjRefMap<T> {
 
     constructor(private readonly config: ObjRefMapConfig<T>) {}
 
-    private idRefToKey(identifier: string, type?: ObjectType): string {
+    private idRefToKey = (identifier: string, type?: ObjectType): string => {
         return !this.config.strictTypeCheck || !type ? identifier : `${identifier}#${type}`;
-    }
+    };
+
+    private shouldInsert = (uri?: string, identifier?: Identifier): boolean => {
+        if (uri && identifier) {
+            const hasUriEntry = this.itemsByUri[uri] !== undefined;
+            const hasIdentifierEntry = this.itemsByIdentifier[identifier] !== undefined;
+
+            // when item has both uri & id, then it is expected that entries either exist or do not exist in _both_
+            //  maps.
+            invariant(hasUriEntry === hasIdentifierEntry);
+
+            return !hasUriEntry;
+        }
 
-    public fromItems(items: ReadonlyArray<T>): ObjRefMap<T> {
+        if (uri) {
+            return this.itemsByUri[uri] === undefined;
+        }
+
+        if (identifier) {
+            return this.itemsByIdentifier[identifier] === undefined;
+        }
+
+        return false;
+    };
+
+    private addItem = (item: T) => {
         const { refExtract, uriExtract, idExtract } = this.config;
 
-        items.forEach((i) => {
-            const uri = uriExtract(i);
+        const uri = uriExtract(item);
+        const identifier = idExtract(item);
 
-            if (this.itemsByUri[uri]) {
-                return;
-            }
+        if (!this.shouldInsert(uri, identifier)) {
+            return;
+        }
+
+        if (uri) {
+            this.itemsByUri[uri] = item;
+        }
+
+        if (identifier) {
+            this.itemsByIdentifier[`${this.idRefToKey(identifier, this.config.type)}`] = item;
```

Comment:
Detail: could be just
```ts
this.itemsByIdentifier[this.idRefToKey(identifier, this.config.type)] = item;
```
right? Or why the template string?

Reply:
fixed